util/topsql/stmtstats/stmtstats.go

```diff
@@ -0,0 +1,224 @@
+// and it is expected that these statistics will eventually be collected and merged
+// in the background.
+type StatementStats struct {
+	mu     sync.Mutex
+	data   StatementStatsMap
+	closed *atomic.Bool
+}
+
+// CreateStatementStats try to create and register an StatementStats.
+// If we are in the initialization phase and have not yet called SetupTopSQL
+// to initialize the top-sql, nothing will happen, and we will get nil. But
+// this scene should never appear, because we always call SetupTopSQL before
+// starting the server, at this moment we cannot receive connections and will
+// not create a valid session. So this case will never happen: "This function
+// returns nil, so this valid session will count nothing".
+func CreateStatementStats() *StatementStats {
+	if manager == nil {
+		return nil
+	}
+	stats := &StatementStats{
+		data:   StatementStatsMap{},
+		closed: atomic.NewBool(false),
+	}
+	manager.register(stats)
+	return stats
+}
+
+// GetOrCreateStatementStatsItem creates the corresponding StatementStatsItem
+// for the specified SQLPlanDigest and timestamp if it does not exist before.
+// GetOrCreateStatementStatsItem is just a helper function, not responsible for
+// concurrency control, so GetOrCreateStatementStatsItem is **not** thread-safe.
+func (s *StatementStats) GetOrCreateStatementStatsItem(sqlDigest, planDigest string, ts int64) *StatementStatsItem {
+	key := SQLPlanDigest{SQLDigest: sqlDigest, PlanDigest: planDigest}
+	tsItem, ok := s.data[key]
+	if !ok {
+		s.data[key] = map[int64]*StatementStatsItem{}
+		tsItem = s.data[key]
+	}
+	item, ok := tsItem[ts]
+	if !ok {
+		tsItem[ts] = NewStatementStatsItem()
+		item = tsItem[ts]
+	}
+	return item
+}
+
+// AddExecCount is used to count the number of executions of a certain
+// SQLPlanDigest within a certain timestamp.
+// AddExecCount is thread-safe.
+func (s *StatementStats) AddExecCount(sqlDigest, planDigest string, ts int64, n uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	item := s.GetOrCreateStatementStatsItem(sqlDigest, planDigest, ts)
+	item.ExecCount += n
+}
+
+// AddKvExecCount is used to count the number of executions of a certain
+// SQLPlanDigest for a certain target within a certain timestamp。
+// AddKvExecCount is thread-safe.
+func (s *StatementStats) AddKvExecCount(sqlDigest, planDigest string, ts int64, target string, n uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	item := s.GetOrCreateStatementStatsItem(sqlDigest, planDigest, ts)
+	item.KvStatsItem.KvExecCount[target] += n
+}
+
+// Take removes all existing StatementStatsMap data from StatementStats.
+// Take is thread-safe.
+func (s *StatementStats) Take() StatementStatsMap {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	data := s.data
+	s.data = StatementStatsMap{}
+	return data
+}
+
+// Close marks StatementStats as "closed".
+// The background goroutine will periodically detect whether each
+// StatementStats has been closed, if so, it will be cleaned up.
+func (s *StatementStats) Close() {
```

Comment:
This close semantic is hard to understand from the StatementStats perspective, as there is no resource for it to close. This close is only used for the stats manager. Also `close` does not provide the information about when it should be called. How about something like `SetStatementIsFinished()`?